Speciation of any species
========================================================

Download reference sequence from any source with taxonomy


I did from [silva](http://www.arb-silva.de/search/)
Searching for Lactobacillaceae
[ screen shot ]

add found sequences to cart
download

[ screen shot]

[ screen shot ]

unpack
gunzip -c <tgz_file> | tar -xvf -

#### Analyze sequences, remove those not in the genus are uncultured ... , or ambiguous 'sp.'
###############
library(Biostrings)
gene <- readRNAStringSet("arb-silva.de_2012-11-17_id65695.fasta")
gene <- as(gene,"DNAStringSet")
length(gene)
matched_names <- sapply(strsplit(names(gene),split=" "),"[",1L)
taxonomy <- sapply(sapply(strsplit(names(gene),split=" "),"[",-1L),paste,collapse=" ")
taxonomy <- strsplit(taxonomy,split=";")
# Isolate the 3 known genus
last_anno <- sapply(taxonomy,function(x) rev(x)[1])
## species are normally names genus species isolate 
sp <- grep ("Lactobacillus|Paralactobacillus|Pediococcus|Sharpea",last_anno)
sp <- setdiff(sp,grep("uncultured|sp.",last_anno)) ## remove uncultured or ambiguous
red.gene <- gene[sp]
red.names <- matched_names[sp]
red.taxo <- sapply(taxonomy[sp],paste,collapse=";")
names(red.gene) <- red.names
last_anno <- last_anno[sp]
genus <- sapply(strsplit(last_anno,split=" "),"[",1L)
species <- sapply(strsplit(last_anno,split=" "),"[",2L)
strain <- sapply(strsplit(last_anno,split=" "),"[",3L)
Lactobacillaceae.taxonomy <- data.frame(ID=red.names,genus=genus,species=species,strain=strain,full.taxonomy=red.taxo)
### only keep full taxonomy
trem <- which (is.na(Lactobacillaceae.taxonomy$genus) | is.na(Lactobacillaceae.taxonomy$species))
Lactobacillaceae.taxonomy <- Lactobacillaceae.taxonomy[-trem,]
write.table(Lactobacillaceae.taxonomy, "Lactobacillaceae.taxonomy",sep="\t",row.names=F,col.names=T,quote=F)
red.gene <- red.gene[-trem]
length(red.gene)
writeXStringSet(red.gene,"Lactobacillaceae.fa")

#### cluster at 100% for redundance

nohup cdhit-est -M 1400 -T 8 -d 0 -c 1.000 -n 9 -i Lactobacillaceae.fa -o Lactobacillaceae.reduced.fa  &

#### look at clustering results
red.gene <- readDNAStringSet("Lactobacillaceae.fa")
Lactobacillaceae.taxonomy <- read.table("Lactobacillaceae.taxonomy",sep="\t",header=T)
cdhit_seq <- readDNAStringSet("Lactobacillaceae.reduced.fa")
cdhit_cluster <- readLines("Lactobacillaceae.reduced.fa.clstr")
clust <- grep("^>Cluster",c(cdhit_cluster,">Cluster"))
cdhit_cluster <- paste(cdhit_cluster,"C",rep(seq.int(1,length(diff(clust))),times=diff(clust)),sep=" ")
cdhit_cluster <- cdhit_cluster[-clust]
test_split <- strsplit(cdhit_cluster,split="\t|nt, >|\\.\\.\\. at +/|\\.\\.\\. |%C |C ")
cluster_mat <- data.frame(matrix(unlist(test_split),ncol=5,byrow=T),stringsAsFactors=F)
cluster_mat$X4 <- as.numeric(gsub("at [+]\\/|% ","",cluster_mat$X4))
cluster_mat$errors <- as.numeric(cluster_mat$X2) - as.numeric(cluster_mat$X2)*(cluster_mat$X4/100)

cluster_mat[is.na(cluster_mat$errors),"errors"] <- 0
cluster_mat <- cluster_mat[order(as.numeric(cluster_mat$X5),-is.na(cluster_mat$X4)),]
colnames(cluster_mat) <- c("ord","Len","ID","Identity","Cluster_ID","Error")
cluster_mat <- cluster_mat[match(names(red.gene),cluster_mat$ID),]

cluster_mat <- data.frame(cluster_mat,Lactobacillaceae.taxonomy[-1])
spXcl<- split(cluster_mat$species, cluster_mat$Cluster_ID)
spXcl.len <- sapply(spXcl,function(x) length(unique(x)))

cluster_names <- sapply(sapply(spXcl,function(x) table(as.character(x))),function(x) paste(paste(names(x),x,sep="."),collapse=";"))
cluster_mat_rep <- cluster_mat[is.na(cluster_mat$Identity),]

cluster_mat_rep$sequence_pool[match(names(cluster_names),cluster_mat_rep$Cluster_ID)] <- cluster_names
rep.gene <- red.gene[match(cluster_mat_rep$ID,names(red.gene))]

length(rep.gene)
writeXStringSet(rep.gene,"Lactobacillaceae.red.fa")
write.table(cluster_mat_rep,"Lactobacillaceae.red.taxonomy",sep="\t",row.names=F,col.names=T,quote=F)

#### then mothur to align, filter
mothur "#align.seqs(candidate=Lactobacillaceae.red.fa, template=/mnt/home/msettles/projects/Forney/Bacterial_16S/Alignment_db/silva.bacteria.fasta, flip=T, processors=12); filter.seqs(fasta=Lactobacillaceae.red.align, processors=12);" 

mothur "#align.seqs(candidate=Lactobacillaceae.red.fa, template=silva.bacteria.fasta, flip=T, processors=12); filter.seqs(fasta=Lactobacillaceae.red.align, processors=12);" 

Lact.align <- read.table("Lactobacillaceae.red.align.report",sep="\t",header=T,as.is=T)
source("../R/getGenBank.R")
gb <- get.GenBank(unique(Lact.align$TemplateName))
gb_mapped <- data.frame(ID=names(gb),genus=sapply(strsplit(attr(gb, "species"),split="_"),"[[",1L),species=sapply(strsplit(attr(gb, "species"),split="_"),"[[",2L))

Lact.align <- data.frame(Lact.align,gb_mapped[match(Lact.align$TemplateName,gb_mapped$ID),])

Lactobacillaceae.red.taxonomy <- read.table("Lactobacillaceae.red.taxonomy",sep="\t",header=T,as.is=F)

keep <- (Lact.align$QueryEnd-Lact.align$QueryLength == 0 & Lact.align$QueryStart==1) & (Lact.align$TemplateEnd >600 & Lact.align$TemplateStart < 50)

#### use mothur to produce distance matrix
mothur "#dist.seqs(fasta=Lactobacillaceae.red.filter.fasta, calc=onegap, output=square, processors=12)"

library(flashClust)
library(WGCNA)
d5k <- read.table("Lactobacillaceae.red.filter.square.dist",skip=1,row.names=1)
#TOM = TOMsimilarity(as.matrix(d5k));
#dissTOM = 1-TOM

hc <- flashClust(as.dist(d5k),method="average")

minModuleSize = 5;
# Module identification using dynamic tree cut:
dynamicMods = cutreeDynamic(dendro = hc, distM = as.matrix(d5k),
 deepSplit = 2, pamRespectsDendro = FALSE,
 minClusterSize = minModuleSize);

dynamicColors = labels2colors(dynamicMods)
speciesColors = labels2colors(cluster_mat_rep$species)
table(dynamicColors)
# Plot the dendrogram and colors underneath
plotDendroAndColors(geneTree, data.frame(dynamicColors,speciesColors), "Dynamic Tree Cut",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main = "Gene dendrogram and module colors")

