```{r, echo=FALSE, include=FALSE}
library("rSFFreader")
```

Pyrosequencing PreProcessing Report File for `r basefilename`  -- Microbial Amplicons
========================================================

Summary
--------------------------------------------------------
--------------------------------------------------------

Raw unclipped Roche 454 pyrosequence reads are cleaned, taxonomically assigned and filtered in
the following manner. Raw SFF files are read directly into the R statistical programming
language using the R package rSFFreader (http://bioconductor.org/packages/2.11/bioc/html/rSFFreader.html, version `r packageVersion("rSFFreader")`), Roche quality clip points are identified and recorded; however, full sequence
reads (unclipped) are used for the identification of Roche 454 adapters, barcodes and amplicon
primers sequence using Cross Match (version `r cross_match_ver`, parameters: min matches=`r cross_match_minmatch`, min
score=`r cross_match_minscore` and -tags) from the phred/phrap/consed application suite. Cross Match alignment information
is then read into R and processed to identify alignment quality, directionality, barcode
assignment, and new read clip points. Base quality clipping was performed using the application
Lucy (version `r Lucy_ver`, parameters: max average error=`r Lucy_max_avg_error`, max error at ends=
`r Lucy_max_error_at_ends`, minimum=0). We then align the clipped reads to the silva bacterial sequence database
using mothur (version `r mothur_ver`). Alignment end points are identified and used in subsequent filtering. Sequence reads
were filtered to only those that met the following criteria: 
* (a) sequence is at least `r minlength`bp in length and no greater than `r maxlength`bp in length; 
* (b) max hamming distance of barcode = `r maxhammingdisttag`; 
* (c) maximum number of matching error to forward primer sequences = `r maxforwardprimererrors`; 
* (d) had < `r maxNs` ambiguous bases (Ns); - After Lucy clipping however, no Ns occur
* (e) had < `r maxhomopol`bp homopolymer run in sequence; 
* (f) alignment to the silva bacterial database is within 75bp of the expected alignment start position (507bp); and 
* (g) read alignment started within the first 5bp and extended through read to within the final 5bp. 
The RDP Bayesian classifier is used to assign sequences to phylotypes (RDP `r rdp_ver`). 
Reads are assigned to the first RDP level with a bootstrap score >=50.

Adapter, Barcode and Primer Identification
--------------------------------------------------------
--------------------------------------------------------

### Basic Read Information
Total Number of Reads in SFF file:`r length(fq)`

Mean length before Roche Right Clip: `r signif(mean(ReadData$RawLength),3)`

Median length before Roche Right Clip: `r signif(median(ReadData$RawLength),3)`

Mean length after Roche Right Clip: `r signif(mean(ReadData$RocheLength),3)`

Median length after Roche Right Clip: `r signif(median(ReadData$RocheLength),3)`

```{r echo=FALSE, fig.align='center',fig.width=9,fig.height=4}
hist(ReadData$RawLength, breaks=200, xlab="Read Length", main="Raw Length")    
hist(ReadData$RocheLength, breaks=200, xlab="Read Length", main="Roche Clipped Length")    
```

### Primer identification using Cross_Match and primer sequences
```{r echo=FALSE}
cross_match_call
```

```{r echo=FALSE, fig.align='center',fig.width=9,fig.height=4}
plot(cm_out$read_start[cm_out$FC=="F"],cm_out$score[cm_out$FC=="F"], main="Foward Match",xlab="basepair position",ylab="score")
plot(cm_out$read_start[cm_out$FC=="C"],cm_out$score[cm_out$FC=="C"], main="Reverse Match", xlab="basepair position",ylab="score")
```


```{r echo=FALSE, fig.align='center',fig.width=9,fig.height=4}
barplot(table(ReadData$Barcode_Err,useNA="always")/length(fq)*100,ylim=c(0,100),main="Barcode Error",xlab="Number of errors",ylab="Percent of reads")
barplot(table(ReadData$FP_Err,useNA="always")/length(fq)*100,ylim=c(0,100),main="5' Primer Error",xlab="Number of errors",ylab="Percent of reads")
```

3' Primers found 
```{r echo=FALSE}
as.data.frame(table(ReadData$Primer_3prime,useNA="always"))
```

Reads with Barcode found : `r table(!is.na(ReadData$Barcode))[2]`

Total Forward Primer Errors <= `r maxforwardprimererrors` :: Reads meeting that criteria: `r table(ReadData$FP_Err <= maxforwardprimererrors)[2]`

Tag Max Hamming Distance from Target <= `r maxhammingdisttag` :: Reads meeting that criteria: `r table(ReadData$Barcode_Err <= maxhammingdisttag)[2]`


### Trimming low quality sequence using Lucy
--------------------------------------------------------
--------------------------------------------------------
```{r echo=FALSE}
Lucy_call
```

```{r echo=FALSE, fig.align='center',fig.width=9,fig.height=4}
hist(ReadData$LucyLC - ReadData$AdapterLC,breaks=200,main="Histogram of Lucy trimming (5' end)", xlab="Number of bases trimmed")
hist(ReadData$AdapterRC - ReadData$LucyRC,breaks=200,main="Histogram of Lucy trimming (3' end)", xlab="Number of bases trimmed")
```

Mean number of bases removed by Lucy clipping: `r signif(mean(ReadData$AdapterLength-ReadData$LucyLength),3)`

Median number of bases removed by Lucy clipping: `r signif(median(ReadData$AdapterLength-ReadData$LucyLength),3)`

Reads Greater than Min Length (`r minlength`) :: Reads meeting that criteria: `table(ReadData$LucyLength > minlength)[2]`

Reads Less than Max Length (`r maxlength`) :: Reads meeting that criteria: `r table(ReadData$LucyLength < maxlength)[2]`

Maximum Number of Abiguous Characters (Ns) <= `r maxNs` :: Reads meeting that criteria: `r table(ReadData$LucyNs <= maxNs)[2]`

Maximum Length of a Homopolymer Run <= `r maxhomopol` :: Reads meeting that criteria: `r table(ReadData$LucymHomoPrun <= maxhomopol)[2]`

Total number of unique reads: `r lucy.unique`

